libraries/chain_kv/include/b1/session/session.hpp

```diff
@@ -842,92 +929,300 @@ const std::shared_ptr<const cache_data_store>& session<persistent_data_store, ca
     return m_impl->cache;
 }
 
+
+template <typename persistent_data_store, typename cache_data_store, typename iterator_traits>
+using session_iterator_alias = typename session<persistent_data_store, cache_data_store>::template session_iterator<iterator_traits>;
+
 template <typename persistent_data_store, typename cache_data_store>
 template <typename iterator_traits>
-const key_value& session<persistent_data_store, cache_data_store>::session_iterator<iterator_traits>::read_(const bytes& key) const {
-    auto* current = m_cache_iterator_state.child.get();
-    while (current) {
-        if (current->cache.contains(key)) {
-            return current->cache.read(key);
+session<persistent_data_store, cache_data_store>::session_iterator<iterator_traits>::session_iterator(const session_iterator& it) 
```

Comment:
I think these copy and move constructors are a bit too computation-heavy. Would it be possible to use a pointer to some data structure, like shared_ptr or something? In that case, copy and move constructors would just copy or move that pointer.

Reply:
The queues maintain a pointer into the iterator states container (which holds stack based instances).  So moving the queues and the iterator states would result in a new iterator states container and the data in the queues pointing into the old iterator states container which means they would be pointing to reclaimed memory.  It's a similar issue with copying, except the queues will be pointing back to the original container instead of the new one.  

The iterator state struct isn't that heavy to copy or move around so it should be fine not be stored on the heap somewhere.